### PR TITLE
[resolves #1889] Cannot obtain direct consumer with spring context

### LINKED ISSUE #1889: Cannot obtain direct consumer with spring context
With 

```
mvn clean install -pl itests/standalone/smoke -am -Dtest=SpringBeanTransformTest
```

I see 

```
09:14:54,402 SEVERE [org.jboss.arquillian.protocol.jmx.JMXTestRunner] (pool-3-thread-1) Failed: org.wildfly.camel.test.spring.SpringBeanTransformTest.testSpringContextFromURL: org.apache.camel.CamelExecutionException: Exception occurred during execution on the exchange: Exchange[ID-localhost-54806-1496042093765-0-2]
	at org.apache.camel.util.ObjectHelper.wrapCamelExecutionException(ObjectHelper.java:1790)
	at org.apache.camel.util.ExchangeHelper.extractResultBody(ExchangeHelper.java:695)
	at org.apache.camel.impl.DefaultProducerTemplate.extractResultBody(DefaultProducerTemplate.java:515)
	at org.apache.camel.impl.DefaultProducerTemplate.extractResultBody(DefaultProducerTemplate.java:511)
	at org.apache.camel.impl.DefaultProducerTemplate.requestBody(DefaultProducerTemplate.java:338)
	at org.wildfly.camel.test.spring.SpringBeanTransformTest.testSpringContextFromURL(SpringBeanTransformTest.java:66)
...
Caused by: org.apache.camel.component.direct.DirectConsumerNotAvailableException: No consumers available on endpoint: direct://start. Exchange[ID-localhost-54806-1496042093765-0-2]
	at org.apache.camel.component.direct.DirectProducer.process(DirectProducer.java:55)
	at org.apache.camel.processor.SharedCamelInternalProcessor.process(SharedCamelInternalProcessor.java:183)
	at org.apache.camel.processor.SharedCamelInternalProcessor.process(SharedCamelInternalProcessor.java:86)
	at org.apache.camel.impl.ProducerCache$1.doInProducer(ProducerCache.java:540)
	at org.apache.camel.impl.ProducerCache$1.doInProducer(ProducerCache.java:505)
	at org.apache.camel.impl.ProducerCache.doInProducer(ProducerCache.java:367)
	at org.apache.camel.impl.ProducerCache.sendExchange(ProducerCache.java:505)
	at org.apache.camel.impl.ProducerCache.send(ProducerCache.java:262)
	at org.apache.camel.impl.DefaultProducerTemplate.requestBody(DefaultProducerTemplate.java:337)
```

on branch [camel-20.x
](https://github.com/wildfly-extras/wildfly-camel/tree/camel-2.20.x)

### PR DESCRIPTION
**Pull request by zregvart**
The meaning of `noStart` flag on `SpringCamelContext` was slightly changed with [CAMEL-11261](https://issues.apache.org/jira/browse/CAMEL-11261), it blocks any `SpringCamelContext` startup even the one invoked explicitly via `start` method.

This changes the creation of SpringCamelContext so that the `noStart` is set and maintained while Spring application context is refreshed, but reset to true after.

This might be considered as a workaround, as the behaviour changed in upstream. I would very much like to remove the `noStart` flag as it is global to all `SpringCamelContext`'s and prefer to use Camel's `isAutoStartup` flag or introduce a new per Camel Context flag. This is perhaps the discussion we need to have in Camel.

### CODE REVIEW

**Comment by tdiesler** [comment]
It doesn't seem to be correct to mandate this sort of change on every client that uses [SpringCamelContextFactory](https://github.com/wildfly-extras/wildfly-camel/blob/master/subsystem/core/src/main/java/org/wildfly/extension/camel/SpringCamelContextFactory.java).

The contract is, that this simple sequence works

```
        CamelContext camelctx = SpringCamelContextFactory.createSingleCamelContext(resourceUrl, null);
        Assert.assertEquals(ServiceStatus.Stopped, camelctx.getStatus());
        camelctx.start();
        try {
            ProducerTemplate producer = camelctx.createProducerTemplate();
            String result = producer.requestBody("direct:start", "Kermit", String.class);
            Assert.assertEquals("Hello green Frog", result);
        } finally {
            camelctx.stop();
        }
```

#1 After the SpringCamelContext is created, it is in state 'stopped'
#2 SpringCamelContext.start() is a blocking operation which makes resources on the context available
#3 SpringCamelContext.start() is also a blocking operation

Hopefully we can keep this contract such that not every client (on the planet) of this API needs to be changed. This API is however an abstraction of the respective Camel & Spring APIs. It means, that we can do all sorts of voodoo [here](https://github.com/wildfly-extras/wildfly-camel/blob/master/subsystem/core/src/main/java/org/wildfly/extension/camel/SpringCamelContextFactory.java#L93).

**Comment by zregvart** [comment]
Yeah, those tests remained from another solution that I tried, rolled them back, the contract should stay the same.

**Comment by zregvart** [comment]
@tdiesler any thoughts on the revised PR? Do you think it would be better to make the change on Camel's side, and keep the previous behaviour?

**Comment by tdiesler** [comment]
The SpringCamelContextFactory creates a list of SpringCamelContext objects from the descriptor. Although I haven't looked at the details of interaction of this ThreadLocal flag and how it affects behavior on the individual instances. From how this is scoped it seems that every SpringCamelContext that is associated with the same Thread is affected.

We are starting/stopping the CamelContext as part of the deployment lifecycle in [CamelContextActivationProcessor|https://github.com/wildfly-extras/wildfly-camel/blob/master/subsystem/core/src/main/java/org/wildfly/extension/camel/deployment/CamelContextActivationProcessor.java]. Because of that we did not want the SpringCamelContext to start automatically. We need this level of control, because previous (pluggable) DeploymentUnitProcessors (DUP) might do additional stuff before this later DUP activates the CamelContext.

In short, CamelContexts being created by SpringCamelContextFactory must not start automatically.

What does this line actually do?

{code}
SpringCamelContext.setNoStart(false)
{code}

If it is a flag that simply allows a later manual start, which would otherwise be prevented, we can accept this PR. If you can handle this at the Camel level without affecting this factory code, even better.

**Comment by zregvart** [comment]
> What does this line actually do?
> 
> SpringCamelContext.setNoStart(false)
> 
> If it is a flag that simply allows a later manual start, which would otherwise be prevented, we can accept this PR. If you can handle this at the Camel level without affecting this factory code, even better.

As it stands now it allows `SpringCamelContext` to start, and that is a change from previous behaviour, previously setting `SpringCamelContext.setNoStart(true)` would prevents automatic startup with the Spring application context, now it prevents any startup. I'm trying to refactor [11261](https://issues.apache.org/jira/browse/CAMEL-11261) to have the new behaviour of automatically starting last and automatically stopping first and have the old behaviour of `noStart` flag. This is proving to be somewhat difficult.

**Comment by zregvart** [comment]
This has been improved by @tdiesler and commited to `camel-2.20.x` branch
